net/redisclient.go

```diff
@@ -226,11 +249,43 @@ func NewRedisRingClient(ro *RedisOptions) *RedisRingClient {
 		r.ring = redis.NewRing(ringOptions)
 		r.log = ro.Log
 		r.metricsPrefix = ro.MetricsPrefix
+
+		if ro.AddrUpdater != nil {
+			if ro.UpdateInterval == 0 {
+				ro.UpdateInterval = defaultUpdateInterval
+			}
+			go func() {
+				old := len(ringOptions.Addrs)
+				t := time.NewTicker(ro.UpdateInterval)
+				r.log.Info("Start goroutine to update redis instances every %s", ro.UpdateInterval)
+				for range t.C {
+					select {
+					case <-r.quit:
+						return
+					default:
+					}
+					addrs := ro.AddrUpdater()
+					if old != len(addrs) && len(addrs) != 0 {
```

Comment:
I think it is safe to call `r.SetAddrs` with the unchanged set of addresses so there is no need to check lengths. Also a single address can change so this check will skip an update.

Reply:
no it would Lock unnecessary and create memory garbage, better filter early

Reply:
> Also a single address can change so this check will skip an update.

What if single redis shard restarts - the len will be the same but address will change

Reply:
> it would Lock unnecessary and create memory garbage

I dont think it is a big problem, one of the goals of https://github.com/go-redis/redis/pull/2093 IMO is to make `SetAddrs` idempotent to make it easy to use on the caller side (i.e. here).

Reply:
Regarding restart you are right I will check the whole list instead of only the number. 

I think we already see problems with linear scalability with number of CPUs in one node and all Locks in the hot path will have something to do with it (my rule of thumb). So in this case I think we can and should be smarter.

Reply:
fixed the restart case